src/Faithlife.Data/DbConnectorCommand.cs

```diff
@@ -345,18 +342,68 @@ string Replacement(Match match)
 					commandText = Regex.Replace(commandText, $@"([?@:]{Regex.Escape(name)})\.\.\.",
 						Replacement, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
 
-					// if special syntax wasn't found, just add the parameter, for databases that support collections directly
-					if (itemCount == -1)
-						AddCommandParameter(name, value);
+					// if special syntax wasn't found, leave the parameter alone, for databases that support collections directly
+					if (itemCount != -1)
+					{
+						parameters = DbParameters.Create(parameters.Take(index).Concat(replacements).Concat(parameters.Skip(index + 1)));
+						index += replacements.Count;
+					}
+					else
+					{
+						index += 1;
+					}
 				}
 				else
 				{
-					AddCommandParameter(name, value);
+					index += 1;
 				}
 			}
 
-			command.CommandText = commandText;
+			IDbCommand command;
+			var transaction = Connector.Transaction;
+
+			var preparedCache = IsPrepared ? Connector.PreparedCommandCache : null;
+			if (preparedCache != null)
+			{
+				if (preparedCache.TryGetValue(commandText, out command))
+				{
+					command.Parameters.Clear();
+					command.Transaction = transaction;
+				}
+				else
+				{
+					command = new PreparedCommand(CreateNewCommand());
+					preparedCache.Add(commandText, command);
+				}
+			}
+			else
+			{
+				command = CreateNewCommand();
+			}
+
+			foreach (var (name, value) in parameters)
+			{
+				if (!(value is IDbDataParameter dbParameter))
+				{
+					dbParameter = command.CreateParameter();
+					dbParameter.Value = value ?? DBNull.Value;
+				}
+
+				dbParameter.ParameterName = name;
+
+				command.Parameters.Add(dbParameter);
+			}
+
 			return command;
```

Comment:
I missed the part where you actually called `command.Prepare()`.

Reply:
LOL, true.